File: library/src/main/java/org/crispy_fish/datatype/underscore_pairs/SimpleStringUnderscorePairReader.java
package org.crispy_fish.datatype.underscore_pairs;

import com.google.common.base.Preconditions;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

public class SimpleStringUnderscorePairReader implements UnderscorePairReader<String> {

    public String get(@Nonnull String pairs, @Nullable String key) {
        Preconditions.checkNotNull(pairs, "pairs must not be null");

        final int keyStartPosition = pairs.indexOf(key);
        if (keyStartPosition < 0) {
            // pairs doesn't contain key
            return null;
        }

        String value;
        final int keyLength = key.length();
        final int valueStartPosition = keyStartPosition + keyLength + 1; // +1 to account for the trailing _
        final int valueStopPosition = pairs.indexOf(Constants.UNDERSCORE, valueStartPosition);
        if (valueStopPosition >= 0) {
            value = pairs.substring(valueStartPosition, valueStopPosition);
        } else {
            value = pairs.substring(valueStartPosition);
        }
        return value;
    }
}

File: library/src/main/kotlin/org/coner/crispyfish/datatype/underscorepairs/SimpleStringUnderscorePairReader.kt
package org.coner.crispyfish.datatype.underscorepairs

class SimpleStringUnderscorePairReader : UnderscorePairReader<String> {

    override fun get(pairs: String, key: String): String? {
        val keyStartPosition = pairs.indexOf(key)
        if (keyStartPosition < 0) {
            // pairs doesn't contain key
            return null
        }

        val keyLength = key.length
        val valueStartPosition = keyStartPosition + keyLength + 1 // +1 to account for the trailing _
        val valueStopPosition = pairs.indexOf('_', valueStartPosition)
        return if (valueStopPosition >= 0) {
            pairs.substring(valueStartPosition, valueStopPosition)
        } else {
            pairs.substring(valueStartPosition)
        }
    }
}
